pkg/pillar/hypervisor/kubevirt.go

```diff
@@ -0,0 +1,1421 @@
+	"kubevirt_vmi_memory_unused_bytes":                 {},
+	"kubevirt_vmi_memory_pgmajfault":                   {},
+	"kubevirt_vmi_memory_pgminfault":                   {},
+	"kubevirt_vmi_memory_swap_in_traffic_bytes_total":  {},
+	"kubevirt_vmi_memory_swap_out_traffic_bytes_total": {},
+}
+
+func newKubevirt() Hypervisor {
+	ctrdCtx, err := initContainerd()
+	if err != nil {
+		logrus.Fatalf("couldn't initialize containerd (this should not happen): %v. Exiting.", err)
+		return nil // it really never returns on account of above
+	}
+
+	switch runtime.GOARCH {
+	case "arm64":
+		return kubevirtContext{
+			ctrdContext:      *ctrdCtx,
+			devicemodel:      "virt",
+			vmiList:          make(map[string]*vmiMetaData),
+			prevDomainMetric: make(map[string]types.DomainMetric),
+		}
+	case "amd64":
+		return kubevirtContext{
+			ctrdContext:      *ctrdCtx,
+			devicemodel:      "pc-q35-3.1",
+			vmiList:          make(map[string]*vmiMetaData),
+			prevDomainMetric: make(map[string]types.DomainMetric),
+		}
+	}
+	return nil
+}
+
+func (ctx kubevirtContext) GetCapabilities() (*types.Capabilities, error) {
+	if ctx.capabilities != nil {
+		return ctx.capabilities, nil
+	}
+	vtd, err := ctx.checkIOVirtualisation()
+	if err != nil {
+		return nil, fmt.Errorf("fail in check IOVirtualization: %v", err)
+	}
+	ctx.capabilities = &types.Capabilities{
+		HWAssistedVirtualization: true,
+		IOVirtualization:         vtd,
+		CPUPinning:               true,
+		UseVHost:                 false, // kubevirt does not support vhost yet
+	}
+	return ctx.capabilities, nil
+}
+
+func (ctx kubevirtContext) checkIOVirtualisation() (bool, error) {
+	f, err := os.Open("/sys/kernel/iommu_groups")
+	if err == nil {
+		files, err := f.Readdirnames(0)
+		if err != nil {
+			return false, err
+		}
+		if len(files) != 0 {
+			return true, nil
+		}
+	}
+	return false, err
+}
+
+func (ctx kubevirtContext) Name() string {
+	return KubevirtHypervisorName
+}
+
+func (ctx kubevirtContext) Task(status *types.DomainStatus) types.Task {
+	return ctx
+}
+
+// Use eve DomainConfig and DomainStatus and generate k3s VMI config or a Pod config
+func (ctx kubevirtContext) Setup(status types.DomainStatus, config types.DomainConfig,
+	aa *types.AssignableAdapters, globalConfig *types.ConfigItemValueMap, file *os.File) error {
+
+	diskStatusList := status.DiskStatusList
+	domainName := status.DomainName
+
+	logrus.Debugf("Setup called for Domain: %s, vmmode %v", domainName, config.VirtualizationMode)
+
+	if config.VirtualizationMode == types.NOHYPER {
+		if err := ctx.CreatePodConfig(domainName, config, status, diskStatusList, aa, file); err != nil {
+			return logError("failed to build kube pod config: %v", err)
+		}
+	} else {
+		// Take eve domain config and convert to VMI config
+		if err := ctx.CreateVMIConfig(domainName, config, status, diskStatusList, aa, file); err != nil {
+			return logError("failed to build kube config: %v", err)
+		}
+	}
+
+	os.MkdirAll(kubevirtStateDir+domainName, 0777)
```

Comment:
Is it really necessary that everybody has access to it?

Reply:
We are trying to figure out what parts of EVE (and pillar) we can run as distinct UID != 0 for some defense in depth. So avoiding creating dirs/files with broad access makes sense.

Reply:
ok changed it to 644 permissions but just FYI. kvm.go writes its state file with 777.